### PR TITLE
Icap rw lock new

### PR DESCRIPTION
**Pull request by chienwei-lan**

### CODE REVIEW

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 1]
```diff
@@ -165,6 +165,15 @@ struct icap {
 	ktime_t			cache_expires;
 
 	enum icap_sec_level	sec_level;
+
+
+	/* Use reader_ref as xclbin metadata reader counter
+	 * Ther reference count increase by 1 if
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c]: if what?

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 2]
```diff
@@ -165,6 +165,15 @@ struct icap {
...
 
 	enum icap_sec_level	sec_level;
+
+
+	/* Use reader_ref as xclbin metadata reader counter
+	 * Ther reference count increase by 1 if 
+	 */
+	u64			busy;
+	int			reader_ref;
+	wait_queue_head_t	ready_wq;
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c]: ready_wq => reader_wq?

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 3]
```diff
@@ -335,6 +344,83 @@ static uint64_t icap_get_data(struct platform_device *pdev, enum data_kind kind)
...
+		return ret;
+
+	mutex_lock(&icap->icap_lock);
+
+	BUG_ON(icap->reader_ref != 0);
+
+done:
+	mutex_unlock(&icap->icap_lock);
+	ICAP_DBG(icap, "%d ret: %d", pid, ret);
+	return ret;
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c]: I'd do:
mutex_lock(&icap->icap_lock);
if (icap->busy)
        ret = -EBUSY;
else
        icap->busy = (u64)pid;
mutex_lock(&icap->icap_lock);
if (ret != 0)
        goto done;

ret = wait_event_interruptible(icap->ready_wq, icap->reader_ref == 0);
if (ret != 0)
        goto done;
BUG_ON(icap->reader_ref != 0);

done:
	ICAP_DBG(icap, "%d ret: %d", pid, ret);
	return ret;

**Comment by chienwei-lan** [comment]
retest this please

**Pull request merged**